aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,349 @@
+        Structure::try_from(type_tag).ok()
+    }};
+}
+
+macro_rules! format_from_ty_arg {
+    ($context:expr, $typ:expr) => {{
+        let type_tag = $context.type_to_type_tag($typ)?;
+        SerializationFormat::try_from(type_tag).ok()
+    }};
+}
+
+macro_rules! store_element {
+    ($context:expr, $obj:expr) => {{
+        let target_vec = &mut $context.extensions_mut().get_mut::<AlgebraContext>().objs;
+        let ret = target_vec.len();
+        target_vec.push(Rc::new($obj));
+        ret
+    }};
+}
+
+macro_rules! abort_unless_feature_flag_enabled {
+    ($context:ident, $flag_opt:expr) => {
+        match $flag_opt {
+            Some(flag) if $context.get_feature_flags().is_enabled(flag) => {
+                // Continue.
+            },
+            _ => {
+                return Err(SafeNativeError::Abort {
+                    abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+                });
+            },
+        }
+    };
+}
+
+macro_rules! abort_unless_single_type_basic_op_enabled {
+    ($context:ident, $structure_opt:expr) => {
+        let flag_opt = feature_flag_of_single_type_basic_op($structure_opt);
+        abort_unless_feature_flag_enabled!($context, flag_opt);
+    };
+}
+
+macro_rules! abort_unless_serialization_format_enabled {
+    ($context:ident, $structure_opt:expr, $format_opt:expr) => {
+        let flag_opt = feature_flag_of_serialization_format($structure_opt, $format_opt);
+        abort_unless_feature_flag_enabled!($context, flag_opt);
+    };
+}
+
+fn abort_invariant_violated() -> PartialVMError {
+    PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+}
+
+/// Try getting a pointer to the `handle`-th elements in `context` and assign it to a local variable `ptr_out`.
+/// Then try casting it to a reference of `typ` and assign it in a local variable `ref_out`.
+/// Abort the VM execution with invariant violation if anything above fails.
+macro_rules! safe_borrow_element {
+    ($context:expr, $handle:expr, $typ:ty, $ptr_out:ident, $ref_out:ident) => {
+        let $ptr_out = $context
+            .extensions()
+            .get::<AlgebraContext>()
+            .objs
+            .get($handle)
+            .ok_or_else(abort_invariant_violated)?
+            .clone();
+        let $ref_out = $ptr_out
+            .downcast_ref::<$typ>()
+            .ok_or_else(abort_invariant_violated)?;
+    };
+}
+
+/// Macros that implements `serialize_internal()` using arkworks libraries.
+macro_rules! ark_serialize_internal {
+    (
+        $gas_params:expr,
+        $context:expr,
+        $args:ident,
+        $structure:expr,
+        $format:expr,
+        $ark_type:ty,
+        $ark_ser_func:ident
+    ) => {{
+        $context.charge($gas_params.placeholder)?;
+        let handle = safely_pop_arg!($args, u64) as usize;
+        safe_borrow_element!($context, handle, $ark_type, element_ptr, element);
+        let mut buf = vec![];
+        element.$ark_ser_func(&mut buf).unwrap();
```

Comment:
This still needs to be checked for error & an invariant violation returned, just like the other one. Nothing to lose if we check. Too much to lose if there is an adversarial input that makes this function return `None`.

Reply:
If serialization failure happens here, it's most likely a memory allocation failure. I'm not sure if in this case aborting the transaction is more correct than panicking - on other node it may execute fine.

Reply:
Fixed.